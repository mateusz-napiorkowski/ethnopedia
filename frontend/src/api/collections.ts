import axios from "axios"
import { Collection } from "../@types/Collection"
import { useMutation } from "react-query"

const API_URL = "http://localhost:8080/api/"
export const getCollections = async () => {
    const response = await axios.get(`${API_URL}v1/collection`)
    return response.data.collections as Collection[]
}

export const useCreateCollectionMutation = () => {
    return useMutation(async (newCollectionData: Collection) => {
        const res = await axios.post(`${API_URL}v1/collection`, newCollectionData)
        return res.data
    });
}

export const getSingleCollection = async (id: string) => {
    const response = await axios.get(`${API_URL}v1/collection/${id}`)
    return response.data.collection as Collection
}
