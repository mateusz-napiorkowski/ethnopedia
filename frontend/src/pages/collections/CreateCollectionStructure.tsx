import Navbar from "../../components/navbar/Navbar"
import React from "react"

const CreateCollectionStructure = () => {



    return (
        <>
            <Navbar />
            <section className="bg-gray-50 dark:bg-gray-900 p-3 sm:p-5 h-full">
                <div className="mx-auto max-w-screen-xl px-4 lg:px-12">
                    CreateCollectionStructure
                </div>
            </section>
        </>
    )
}

export default CreateCollectionStructure