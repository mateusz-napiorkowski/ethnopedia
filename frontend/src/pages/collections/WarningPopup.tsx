import { ReactComponent as CloseIcon } from "../../assets/icons/close.svg"
import React from "react"

type Props = {
    onClose: () => void
    deleteSelected: () => void
    warningMessage: string
}

const WarningPopup = ({ onClose, deleteSelected, warningMessage }: Props) => {
    return <div id="popup-modal"
                className="overflow-y-auto flex items-center overflow-x-hidden fixed z-50 justify-center items-center w-full md:inset-0 h-[calc(100%-1rem)] max-h-full ">
        <div className="fixed inset-0 bg-black opacity-50" />

        <div className="relative p-4 w-full max-w-md max-h-full ">
            <div className="relative bg-white rounded-lg shadow-md dark:bg-gray-700 border-gray-200 border">
                <div className="flex justify-end">
                    <button type="button" className="items-end content-end border-none" aria-label="exit"
                            onClick={onClose}>
                        <CloseIcon />
                    </button>
                </div>

                <hr className="border-t border-gray-200 dark:border-gray-600 w-8 self-center w-full" />

                <div className="p-4 md:p-5 text-center">
                    <h3 className="mb-5 text-lg font-normal text-gray-700 font-semibold dark:text-gray-400">

                        {warningMessage}
                    </h3>
                    <button aria-label="confirm" data-modal-hide="popup-modal" type="button"
                            className="text-white bg-red-600 font-semibold hover:bg-red-800 focus:ring-4
                            focus:outline-none focus:ring-red-600 dark:focus:ring-red-800
                            font-medium rounded-lg text-sm inline-flex items-center px-5 py-2.5 text-center me-2"
                    onClick={deleteSelected}>
                        Usu≈Ñ
                    </button>
                    <button data-modal-hide="popup-modal" type="button" onClick={onClose}
                            className="text-gray-800 bg-white hover:bg-gray-100 focus:ring-4 focus:outline-none
                            font-semibold
                            focus:ring-gray-200 rounded-lg border border-gray-300 text-sm font-medium px-5 py-2.5
                            hover:text-gray-900 focus:z-10 dark:bg-gray-700 dark:text-gray-300 dark:border-gray-500
                            dark:hover:text-white dark:hover:bg-gray-600 dark:focus:ring-gray-600">
                        Anuluj
                    </button>
                </div>
            </div>
        </div>
    </div>
}

export default WarningPopup