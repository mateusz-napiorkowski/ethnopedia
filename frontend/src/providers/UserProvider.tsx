import React, { createContext, useContext, useState, useEffect } from "react"
import { jwtDecode } from "jwt-decode"

export interface JWT {
    exp: number;
    username: string;
    firstName: string;
    userId: string;
}

interface UserContextProps {
    isUserLoggedIn: boolean;
    firstName: string;
    userId: string;
    jwtToken: string;
    setUserData: (isLoggedIn: boolean, name: string, jwtToken: string, userId: string) => void;
}

export const UserContext = createContext<UserContextProps>({
    isUserLoggedIn: false,
    firstName: "",
    userId: "",
    jwtToken: "",
    setUserData: () => {
    },
})

export const useUser = () => useContext(UserContext)

interface UserProviderProps {
    children: React.ReactNode;
}

export const UserProvider: React.FC<UserProviderProps> = ({ children }) => {
    const [isUserLoggedIn, setIsUserLoggedIn] = useState(false)
    const [firstName, setFirstName] = useState("")
    const [userId, setUserId] = useState("")
    const [jwtToken, setJwtToken] = useState("")

    const setUserData = (isLoggedIn: boolean, name: string, jwtToken: string, userId: string) => {
        setIsUserLoggedIn(isLoggedIn)
        setFirstName(name)
        setUserId(userId)
        setJwtToken(jwtToken)
    }

    useEffect(() => {
        const token = localStorage.getItem("token")

        if (token) {
            try {
                const decodedToken = jwtDecode<JWT>(token)
                const currentTime = Date.now() / 1000
                if (decodedToken.exp > currentTime) {
                    setUserData(true, decodedToken.firstName, token, decodedToken.userId)
                } else {
                    localStorage.removeItem("token")
                    setUserData(false, "", "", "")
                }
            } catch (error) {
                console.error("Invalid token:", error)
                localStorage.removeItem("token")
                setUserData(false, "", "", "")
            }
        }
    }, [])

    return (
        <UserContext.Provider value={{ isUserLoggedIn, firstName, userId, jwtToken, setUserData }}>
            {children}
        </UserContext.Provider>
    )
}
