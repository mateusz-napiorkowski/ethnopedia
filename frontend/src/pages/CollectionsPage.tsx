import { useQuery } from "react-query"
import { getCollections } from "../api/collections"
import LoadingPage from "./LoadingPage"
import { Collection } from "../@types/Collection"
import { useNavigate } from "react-router-dom"
import React, { useState } from "react"
import { ReactComponent as PlusIcon } from "../assets/icons/plus.svg"
import CreateCollection from "../components/CreateCollection"

const CollectionsPage = () => {
    const { data: fetchedData } = useQuery(
        ["collection"],
        getCollections,
    )

    const [showPopup, setShowNewCollectionPopup] = useState(false)
    const navigate = useNavigate()

    if (fetchedData === undefined) {
        return <LoadingPage />
    } else {
        const allCollections = fetchedData.map((collection: Collection) => (
            <div className="px-4 py-3 bg-white dark:bg-gray-800 shadow-md rounded-lg mb-4 border dark:border-gray-700
            cursor-pointer"
                 key={collection._id}
                 onClick={() => navigate(`/artworks/search?Kategoria=${collection.name}`)}>

                <div className="flex flex-row">
                    <input className="mr-4" type="checkbox" id="exampleCheckbox" name="exampleCheckbox" />
                    <div>
                        <h2 className="text-lg font-semibold">{collection.name}</h2>
                        <p className="text-gray-600 dark:text-gray-300">{collection.description}</p>
                    </div>
                </div>
            </div>
        ))

        return <section className="bg-gray-50 dark:bg-gray-900 p-3 sm:p-5 h-full">
            {showPopup && <CreateCollection onClose={() => setShowNewCollectionPopup(!showPopup)} />}
            <div className="mx-auto max-w-screen-xl px-4 lg:px-12">
                <button
                    type="button"
                    className="flex items-center justify-center dark:text-white bg-white
                                    hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 font-medium rounded-lg
                                    text-sm px-4 py-2 dark:bg-primary-600 dark:hover:bg-primary-700 focus:outline-none
                                    dark:focus:ring-primary-800 mb-2"
                    onClick={() => setShowNewCollectionPopup(!showPopup)}>
                    <span className="mr-2">
                        <PlusIcon />
                    </span>
                    Dodaj nową kolekcję
                </button>
                {allCollections}
            </div>
        </section>
    }
}
export default CollectionsPage