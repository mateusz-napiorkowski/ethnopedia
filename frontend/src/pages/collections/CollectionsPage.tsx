import CreateCollectionModal from "./CreateCollectionModal"
import SortOptions from "../../components/SortOptions"
import LoadingPage from "../LoadingPage"
import React, { useEffect, useState } from "react"
import WarningPopup from "./WarningPopup"
import { Collection } from "../../@types/Collection"
import { ReactComponent as FileExportIcon } from "../../assets/icons/fileExport.svg"
import { ReactComponent as FileImportIcon } from "../../assets/icons/fileImport.svg"
import { ReactComponent as PlusIcon } from "../../assets/icons/plus.svg"
import { getAllCollections, useBatchDeleteCollectionMutation } from "../../api/collections"
import { useNavigate } from "react-router-dom"
import { useQuery, useQueryClient } from "react-query"
import { useUser } from "../../providers/UserProvider"
import Pagination from "../../components/Pagination"
import { getXlsxWithCollectionData } from "../../api/dataExport"
import ImportOptions from "../../components/ImportOptions"

interface Option {
    value: string
    label: string
}

const CollectionsPage = () => {
    const { firstName } = useUser()
    const [showImportOptions, setShowImportOptions] = useState<boolean>(false)
    const [checkedCollections, setCheckedCollections] = useState<{ [key: string]: boolean }>({})
    const [showWarningPopup, setShowWarningPopup] = useState(false)
    const [currentPage, setCurrentPage] = useState(1)
    const [exportErrorMessage, setExportErrorMessage] = useState("")
    const { jwtToken } = useUser();
    const pageSize = 10

    const queryClient = useQueryClient()

    const { mutate: batchDeleteMutation } = useBatchDeleteCollectionMutation()

    const[newCollection, setNewCollection]=useState<string>("");

    useEffect(() => {
        refetch()
// eslint-disable-next-line
    }, [newCollection]);

    const { data: fetchedData, refetch } = useQuery(
        ["collection", currentPage, pageSize, newCollection],
        () => getAllCollections(currentPage, pageSize),
        {
            keepPreviousData: true,
        },
    )
    const checkAll = () => {
        const newCheckedCollections = fetchedData?.collections?.reduce(
            (acc: any, collection: any) => ({
                ...acc,
                [collection.id!]: true,
            }),
            {},
        ) as Record<string, boolean>

        setCheckedCollections(newCheckedCollections)
    }

    const uncheckAll = () => {
        setCheckedCollections({})
    }

    const handleCheck = (id: string) => {
        setCheckedCollections((prev) => ({ ...prev, [id]: !prev[id] }))
    }

    const deleteSelected = () => {
        const selectedIds = Object.keys(checkedCollections).filter(id => checkedCollections[id])
        if (selectedIds.length > 0) {
            batchDeleteMutation([selectedIds, jwtToken],
                {
                    onSuccess: () => {
                        queryClient.invalidateQueries(["collection"])
                        setShowWarningPopup(!showWarningPopup)
                    },
                })
        }
    }

    const [showPopup, setShowNewCollectionPopup] = useState(false)
    const [sortOrder, setSortOrder] = useState("A-Z")

    const navigate = useNavigate()

    if (fetchedData === undefined) {
        return <LoadingPage />
    } else {
        const sortedCollections = fetchedData.collections ? [...fetchedData.collections].sort((a, b) => {
            if (sortOrder === "A-Z") {
                return a.name.localeCompare(b.name)
            } else {
                return b.name.localeCompare(a.name)
            }
        }) : []

        const allCollections = sortedCollections.map((collection: Collection) => (
            <div
                className="px-4 py-3 bg-white dark:bg-gray-800 shadow-md rounded-lg mb-4 border border-gray-300 dark:border-gray-600 cursor-pointer"
                key={collection.id}
                onClick={() => navigate(`/collections/${collection.name}/artworks`)}
            >

                <div className="flex flex-row justify-between">
                    <div className="flex">
                        <span className="mr-4 items-center flex">
                            <input
                                type="checkbox"
                                checked={checkedCollections[collection.id!] || false}
                                onClick={(e) => e.stopPropagation()}
                                onChange={() => {
                                    handleCheck(collection.id!)
                                }} />
                        </span>

                        <div className="flex-grow">
                            <h2 className="text-lg font-semibold">{collection.name}</h2>
                            <p className="text-gray-600 dark:text-gray-300">{collection.description}</p>
                        </div>
                    </div>

                    {/* <div className="flex flex-col"> */}
                        <h2 className="text-md min-w-fit items-center flex mx-2">
                        <span className="font-bold mr-1">
                            {collection.artworksCount ?? 0}
                        </span>
                            {
                                (collection.artworksCount ?? 0) === 1 ? "rekord" :
                                    (collection.artworksCount ?? 0) > 1 && (collection.artworksCount ?? 0) < 5 ? "rekordy" : "rekordów"
                            }
                        </h2>
                </div>
            </div>
        ))

        const sortOptions: Option[] = [
            { value: "A-Z", label: "Kolekcja rosnąco" },
            { value: "Z-A", label: "Kolekcja malejąco" },
        ]

        return <section className="bg-gray-50 dark:bg-gray-900 p-3 sm:p-5 h-full">
            {showPopup && <CreateCollectionModal stateChanger={setNewCollection} onClose={() => setShowNewCollectionPopup(!showPopup)} />}
            {showWarningPopup && <WarningPopup onClose={() => setShowWarningPopup(!showWarningPopup)}
                                               deleteSelected={deleteSelected}
                                               warningMessage={"Czy na pewno chcesz usunąć zaznaczone kolekcje?"} />}
            <div className="mx-auto max-w-screen-xl px-4 lg:px-12">
                <div className="flex justify-end text-red-500 ml-2 px-1 h-5">
                    <p>{exportErrorMessage}</p>
                </div>
                <div className="flex flex-row">
                    <div className="w-full">
                        <h1 className="font-bold text-4xl mb-4">
                            Witaj{firstName ? ` ${firstName}` : ""}!
                        </h1>
                        <h2 className="mb-2 text-lg">
                            Twoje kolekcje:
                        </h2>
                    </div>

                    <div className="flex items-center justify-end w-full">
                        <button
                            type="button"
                            className="flex items-center justify-center dark:text-white
                                    text-sm px-4 py-2 mb-2 hover:bg-gray-700 bg-gray-800 text-white border-gray-800
                                    font-semibold mr-2"
                            onClick={() => setShowNewCollectionPopup(!showPopup)}>
                            <span className="mr-2">
                                <PlusIcon />
                            </span>
                            Nowa kolekcja
                        </button>

                        <button
                            className="flex items-center justify-center dark:text-white
                                    text-sm px-4 py-2 mb-2 hover:bg-gray-700 bg-gray-800 text-white border-gray-800
                                    font-semibold mr-2"
                            type="button"
                            onClick={() => {
                                if(Object.keys(checkedCollections).length === 1) {
                                    for(const key in fetchedData.collections) {
                                        if(fetchedData.collections[key].id === Object.keys(checkedCollections)[0]) {
                                            getXlsxWithCollectionData(fetchedData.collections[key].name)
                                            setExportErrorMessage("")
                                        }
                                    }
                                } else {
                                    if(Object.keys(checkedCollections).length === 0){
                                        setExportErrorMessage("Najpierw należy zaznaczyć kolekcję do wyeksportowania.")
                                    } else {
                                        const checkedCollectionsNames = Object.keys(checkedCollections)
                                            .filter((checkedCollection) => checkedCollections[checkedCollection])
                                            .map((checkedCollection) => {
                                                const collection = fetchedData.collections.find((col) => col.id == checkedCollection);
                                                return collection!.name;
                                            })
                                        for(const collectionName of checkedCollectionsNames) {
                                            getXlsxWithCollectionData(collectionName)
                                        }
                                        setExportErrorMessage("")
                                    }
                                }
                            }}
                        >
                            <span className="text-white">
                                <FileExportIcon />
                            </span>
                            Eksportuj kolekcję
                        </button>
                        {
                            jwtToken && <button
                            className="flex items-center justify-center dark:text-white
                                    text-sm px-4 py-2 mb-2 hover:bg-gray-700 bg-gray-800 text-white border-gray-800
                                    font-semibold"
                            type="button"
                            onClick={() => setShowImportOptions(showImportOptions => !showImportOptions)}
                        >
                            <FileImportIcon />
                            Importuj kolekcję
                        </button>
                        }
                        {
                            !jwtToken && <button
                            className="flex items-center justify-center dark:text-white
                                    text-sm px-4 py-2 mb-2 hover:bg-gray-600 bg-gray-600 text-white border-gray-800
                                    font-semibold"
                            type="button" disabled={true} title={"Aby zaimportować kolecję musisz się zalogować."}
                            onClick={() => setShowImportOptions(showImportOptions => !showImportOptions)}
                        >
                            <FileImportIcon />
                            Importuj kolekcję
                        </button>
                        }
                        

                    </div>
                </div>
                {showImportOptions && <ImportOptions onClose={() => setShowImportOptions(false)} />}
                <div className="flex flex-row">
                    <div className="flex flex-1">
                        <button type="button" className="px-4 py-2 mb-2 bg-white"
                                onClick={checkAll}>
                            Zaznacz wszystkie
                        </button>

                        <button type="button" className="px-4 py-2 mb-2 ml-2 bg-white"
                                onClick={uncheckAll}>
                            Odznacz wszystkie
                        </button>

                        {jwtToken && <button type="button" className="px-4 py-2 mb-2 ml-2 bg-white"
                                onClick={() => {
                                    if (Object.keys(checkedCollections).length !== 0)
                                        setShowWarningPopup(!showWarningPopup)
                                }}>
                            Usuń zaznaczone
                        </button>}
                        {!jwtToken && <button type="button" disabled={true} title={"Aby usuwać kolecje musisz się zalogować."} className="px-4 py-2 mb-2 ml-2 bg-gray-100 hover:bg-gray-100"
                                onClick={() => {
                                    if (Object.keys(checkedCollections).length !== 0)
                                        setShowWarningPopup(!showWarningPopup)
                                }}>
                            Usuń zaznaczone
                        </button>}
                    </div>
                    <span className="mb-2">
                    <SortOptions
                        options={sortOptions}
                        onSelect={(value: string) => setSortOrder(value)}
                        sortOrder=""
                        setCurrentPage={() => setCurrentPage}
                    />
                    </span>
                </div>

                {allCollections}

                <div className="flex justify-center">
                    <Pagination
                        currentPage={currentPage}
                        totalPages={Math.ceil(fetchedData.total / pageSize)}
                        setCurrentPage={(page) => setCurrentPage(page)}
                    />
                </div>
            </div>
        </section>
    }
}
export default CollectionsPage