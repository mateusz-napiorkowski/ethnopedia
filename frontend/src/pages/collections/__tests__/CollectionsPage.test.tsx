import { fireEvent, render, screen, waitFor } from "@testing-library/react"
import CollectionsPage from "../CollectionsPage";
import { getAllCollections } from "../../../api/collections";
import { QueryClient, QueryClientProvider } from "react-query";
import { MemoryRouter } from "react-router-dom";
import { useUser } from "../../../providers/UserProvider";

jest.mock("../../../api/collections", () => ({
    getAllCollections: jest.fn(),
    useBatchDeleteCollectionMutation: jest.fn(() => ({
        mutate: jest.fn(),
    })),
}));

jest.mock("../../../providers/UserProvider", () => ({
    useUser: jest.fn(),
}));

const mockCollections = {
    collections: [
        { id: "1", name: "Kolekcja A", description: "Opis A", artworksCount: 1 },
        { id: "2", name: "Kolekcja B", description: "Opis B", artworksCount: 3 },
        { id: "3", name: "Kolekcja C", description: "Opis C", artworksCount: 5 },
    ],
    total: 3,
};

describe("CollectionsPage", () => {
    beforeEach(() => {
        (useUser as jest.Mock).mockReturnValue({ firstName: "Jan", jwtToken: "test-token" });
        (getAllCollections as jest.Mock).mockResolvedValue(mockCollections);
    });

    it("renders the correct user greeting and static header", async () => {
        const queryClient = new QueryClient();

        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Sprawdzamy, czy nagłówki z powitaniem i tekstem "Twoje kolekcje" są renderowane
        expect(await screen.findByText("Witaj Jan!")).toBeInTheDocument();
        expect(screen.getByText("Twoje kolekcje:")).toBeInTheDocument();
    });

    it("displays the list of collections correctly", async () => {
        const queryClient = new QueryClient();

        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Czekamy na załadowanie wszystkich kolekcji
        const collectionItems = await screen.findAllByRole('heading', {
            name: /Kolekcja/i
        });
        expect(collectionItems.length).toBe(3);

        // Sprawdzamy każdą kolekcję
        expect(screen.getByText("Kolekcja A")).toBeInTheDocument();
        expect(screen.getByText("Kolekcja B")).toBeInTheDocument();
        expect(screen.getByText("Kolekcja C")).toBeInTheDocument();
    });

    it("displays the correct number of artworks for each collection", async () => {
        const queryClient = new QueryClient();

        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Poczekaj na załadowanie elementów
        await waitFor(() => {
            const elements = screen.getAllByText(/rekord/);
            expect(elements.length).toBe(3);
        });

        // Sprawdź każdy z elementów po załadowaniu
        const elements = screen.getAllByText(/rekord/);

        // Zaktualizowane asercje z usunięciem spacji
        expect(elements[0]).toHaveTextContent("1rekord");
        expect(elements[1]).toHaveTextContent("3rekordy");
        expect(elements[2]).toHaveTextContent("5rekordów");
    });

    it("sorts the collections correctly by A-Z and Z-A", async () => {
        const queryClient = new QueryClient();

        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Sprawdzamy, czy początkowa kolejność kolekcji to A-Z
        let collectionItems = await screen.findAllByRole('heading', {
            name: /Kolekcja/i
        });

        // Sprawdzamy, czy kolekcje są posortowane A-Z
        expect(collectionItems[0]).toHaveTextContent("Kolekcja A");
        expect(collectionItems[1]).toHaveTextContent("Kolekcja B");
        expect(collectionItems[2]).toHaveTextContent("Kolekcja C");

        // Zmieniamy kolejność na Z-A
        const sortDropdown = screen.getByRole("combobox");
        fireEvent.change(sortDropdown, { target: { value: "Z-A" } });

        // Sprawdzamy, czy kolekcje są posortowane Z-A
        collectionItems = await screen.findAllByRole('heading', {
            name: /Kolekcja/i
        });

        expect(collectionItems[0]).toHaveTextContent("Kolekcja C");
        expect(collectionItems[1]).toHaveTextContent("Kolekcja B");
        expect(collectionItems[2]).toHaveTextContent("Kolekcja A");
    });

    it("checks and unchecks collections correctly", async () => {
        const queryClient = new QueryClient();

        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Poczekaj na załadowanie kolekcji, żeby upewnić się, że checkboxes są dostępne
        await waitFor(() => screen.findAllByRole('checkbox'));

        // Sprawdzamy początkowy stan zaznaczenia (wszystkie odznaczone)
        let checkboxes = screen.getAllByRole('checkbox');
        checkboxes.forEach(checkbox => {
            expect(checkbox).not.toBeChecked();
        });

        // Zaznaczamy wszystkie kolekcje
        const selectAllButton = screen.getByText("Zaznacz wszystkie");
        fireEvent.click(selectAllButton);

        // Sprawdzamy, czy wszystkie kolekcje zostały zaznaczone
        checkboxes = screen.getAllByRole('checkbox');
        checkboxes.forEach(checkbox => {
            expect(checkbox).toBeChecked();
        });

        // Odznaczamy wszystkie kolekcje
        const unselectAllButton = screen.getByText("Odznacz wszystkie");
        fireEvent.click(unselectAllButton);

        // Sprawdzamy, czy wszystkie kolekcje zostały odznaczone
        checkboxes = screen.getAllByRole('checkbox');
        checkboxes.forEach(checkbox => {
            expect(checkbox).not.toBeChecked();
        });
    });

    it("opens the 'Nowa kolekcja' modal on button click", async () => {
        const queryClient = new QueryClient();
        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Klikamy przycisk 'Nowa kolekcja'
        const newCollectionButton = await screen.findByRole("button", { name: /Nowa kolekcja/i });
        fireEvent.click(newCollectionButton);

        // Sprawdzamy, czy modal został otwarty
        expect(await screen.findByText("Dodaj nową kolekcję")).toBeInTheDocument();
    });


    it("opens the 'Usuń zaznaczone' warning popup when button is clicked", async () => {
        const queryClient = new QueryClient();
        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Zaznaczamy kolekcje
        const checkboxes = await screen.findAllByRole("checkbox");
        fireEvent.click(checkboxes[0]); // Zaznaczamy pierwszą kolekcję

        // Klikamy przycisk "Usuń zaznaczone"
        const deleteButton = screen.getByText("Usuń zaznaczone");
        fireEvent.click(deleteButton);

        // Sprawdzamy, czy popup został otwarty
        expect(await screen.findByText("Czy na pewno chcesz usunąć zaznaczone kolekcje?")).toBeInTheDocument();
    });


    it("changes the collection sort order when the dropdown is changed", async () => {
        const queryClient = new QueryClient();
        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Sprawdzamy początkową kolejność kolekcji
        let collectionItems = await screen.findAllByRole('heading', { name: /Kolekcja/i });
        expect(collectionItems[0]).toHaveTextContent("Kolekcja A");

        // Zmieniamy kolejność na Z-A
        const sortDropdown = screen.getByRole("combobox");
        fireEvent.change(sortDropdown, { target: { value: "Z-A" } });

        // Sprawdzamy nową kolejność kolekcji
        collectionItems = await screen.findAllByRole('heading', { name: /Kolekcja/i });
        expect(collectionItems[0]).toHaveTextContent("Kolekcja C");
    });

    it("displays an error message when no collections are selected for export", async () => {
        const queryClient = new QueryClient();
        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Czekamy na pojawienie się przycisku eksportu
        const exportButton = await screen.findByRole("button", { name: /Eksportuj kolekcję/i });
        fireEvent.click(exportButton);

        // Sprawdzamy, czy pojawił się komunikat o błędzie
        expect(await screen.findByText("Najpierw należy zaznaczyć kolekcję do wyeksportowania.")).toBeInTheDocument();
    });

    it("disables the import button when the user is not logged in", async () => {
        (useUser as jest.Mock).mockReturnValue({ firstName: "Jan", jwtToken: null }); // Mocking no jwtToken

        const queryClient = new QueryClient();
        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Sprawdzamy, czy przycisk "Importuj kolekcję" jest wyłączony
        const importButton = await screen.findByText("Importuj kolekcję");
        expect(importButton).toBeDisabled();
    });

    it("shows an error message when no collection is selected for export", async () => {
        const queryClient = new QueryClient();
        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter>
                    <CollectionsPage />
                </MemoryRouter>
            </QueryClientProvider>
        );

        // Klikamy przycisk eksportu bez zaznaczenia kolekcji
        const exportButton = await screen.findByText("Eksportuj kolekcję");
        fireEvent.click(exportButton);

        // Sprawdzamy, czy wyświetla się komunikat o błędzie
        expect(await screen.findByText("Najpierw należy zaznaczyć kolekcję do wyeksportowania.")).toBeInTheDocument();
    });

    // it("paginates through collections correctly", async () => {
    //     const mockCollectionsPage1 = {
    //         collections: [
    //             { id: "1", name: "Kolekcja A", description: "Opis A", artworksCount: 1 },
    //             { id: "2", name: "Kolekcja B", description: "Opis B", artworksCount: 3 },
    //         ],
    //         total: 5,
    //     };
    //
    //     const mockCollectionsPage2 = {
    //         collections: [
    //             { id: "3", name: "Kolekcja C", description: "Opis C", artworksCount: 5 },
    //             { id: "4", name: "Kolekcja D", description: "Opis D", artworksCount: 2 },
    //         ],
    //         total: 5,
    //     };
    //
    //     (getAllCollections as jest.Mock)
    //         .mockResolvedValueOnce(mockCollectionsPage1)
    //         .mockResolvedValueOnce(mockCollectionsPage2);
    //
    //     const queryClient = new QueryClient();
    //
    //     render(
    //         <QueryClientProvider client={queryClient}>
    //             <MemoryRouter>
    //                 <CollectionsPage />
    //             </MemoryRouter>
    //         </QueryClientProvider>
    //     );
    //
    //     // Check the first page
    //     let collectionItems = await screen.findAllByRole("heading", { name: /Kolekcja/i });
    //     expect(collectionItems.length).toBe(2); // Page 1 should show 2 collections
    //     expect(screen.getByText("Kolekcja A")).toBeInTheDocument();
    //     expect(screen.getByText("Kolekcja B")).toBeInTheDocument();
    //
    //     // Click next page button
    //     const nextPageButton = screen.getByText("Next");
    //     fireEvent.click(nextPageButton);
    //
    //     // Wait for new collections
    //     collectionItems = await screen.findAllByRole("heading", { name: /Kolekcja/i });
    //     expect(collectionItems.length).toBe(2); // Page 2 should show 2 collections
    //     expect(await screen.findByText("Kolekcja C")).toBeInTheDocument();
    //     expect(await screen.findByText("Kolekcja D")).toBeInTheDocument();
    //
    //     // Click previous page button
    //     const prevPageButton = screen.getByText("Previous");
    //     fireEvent.click(prevPageButton);
    //
    //     // Wait for the first page collections to load again
    //     collectionItems = await screen.findAllByRole("heading", { name: /Kolekcja/i });
    //     expect(collectionItems.length).toBe(2); // Back to page 1
    //     expect(screen.getByText("Kolekcja A")).toBeInTheDocument();
    //     expect(screen.getByText("Kolekcja B")).toBeInTheDocument();
    // });
});
