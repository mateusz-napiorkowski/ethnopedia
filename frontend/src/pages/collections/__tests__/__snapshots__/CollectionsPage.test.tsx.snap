// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionsPage tests renders loading state 1`] = `
<div>
  <div
    class="flex justify-center items-center h-screen bg-gray-50 dark:bg-gray-900"
    role="status"
  >
    <svg />
  </div>
</div>
`;
